### PR TITLE
Add history and health.

### PR DESCRIPTION
**Pull request by ancabusuioc**
Added health and history endpoints to PAM - https://jira.ft.com/browse/UPP-1177

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -69,8 +75,14 @@ var errorLogger *log.Logger
 var configFileName = flag.String("config", "", "Path to configuration file")
 var appConfig *AppConfig
 var metricSink = make(chan PublishMetric)
+var metricContainer publishHistory
+
+
+var cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")
```
Comment (tamas-molnar) [inline app.go]: - unused, also commented code on line 84 - maybe add a config to enable cpu profiling?
- rename to camelCase

File: app.go [hunk 2]
```diff
@@ -126,6 +141,24 @@ func startAggregator() {
 	go aggregator.Run()
 }
 
+func loadHistory(w http.ResponseWriter, r *http.Request) {
+	log.Printf("History request.")
+	infoLogger.Printf("size:%s",len(metricContainer.publishMetrics))
+	metricContainer.RLock()
+	for i := len(metricContainer.publishMetrics) - 1; i >= 0; i-- {
+		fmt.Fprintf(w, "%d. Tid: %s, UUID: %s, Endpoint: %s, PublishDate: %s, Duration: %d, Succeeded: %t\n\n",
```
Comment (tamas-molnar) [inline app.go]: - Maybe implement this as a String() method on PublishMetric
- should we make it CSV? Who will use this output?

File: healthcheck.go [hunk 3]
```diff
@@ -107,3 +108,31 @@ func checkIfTopicIsPresent(body []byte, searchedTopic string) error {
...
+	metricContainer.RLock()
+	failures := 0;
+	for i := 0; i < len(metricContainer.publishMetrics); i++ {
+		if (!metricContainer.publishMetrics[i].publishOK) {
+			failures += 1
+		}
+	}
+	metricContainer.RUnlock()
+
+	if failures >= 2 {
```
Comment (tamas-molnar) [inline healthcheck.go]: Should this be configurable?

File: app.go [hunk 4]
```diff
@@ -88,10 +102,11 @@ func main() {
 
 func enableHealthchecks() {
 
-	healthcheck := &Healthcheck{http.Client{}, *appConfig}
+	healthcheck := &Healthcheck{http.Client{}, *appConfig, &metricContainer}
 	router := mux.NewRouter()
 	router.HandleFunc("/__health", healthcheck.checkHealth())
 	router.HandleFunc("/__gtg", healthcheck.gtg)
+	router.HandleFunc("/history", loadHistory)
```
Comment (tamas-molnar) [inline app.go]: Why not `__history`?

File: app.go [hunk 5]
```diff
@@ -81,6 +93,8 @@ func main() {
 		return
 	}
 
+	metricContainer = publishHistory{sync.RWMutex{}, make([]PublishMetric, 0)}
```
Comment (tamas-molnar) [inline app.go]: Should this be limited in size maybe?

**Comment by tamas-molnar** [comment]
Please run `golint`, `govet` against the project.

**Comment by Zapadlo** [comment]
@tamas-molnar Lays down the harshest judgement, its hard to follow :) +1

File: app.go [hunk 6]
```diff
@@ -126,6 +152,16 @@ func startAggregator() {
 	go aggregator.Run()
 }
 
+func loadHistory(w http.ResponseWriter, r *http.Request) {
+	log.Printf("History request.")
+	infoLogger.Printf("size:%d", len(metricContainer.publishMetrics))
```
Comment (BerniVarga) [inline app.go]: I think these log prints can be removed.

File: app.go [hunk 7]
```diff
@@ -99,6 +118,13 @@ func enableHealthchecks() {
 	}
 }
 
+func attachProfiler(router *mux.Router) {
+	router.HandleFunc("/debug/pprof/", pprof.Index)
+	router.HandleFunc("/debug/pprof/cmdline", pprof.Cmdline)
+	router.HandleFunc("/debug/pprof/profile", pprof.Profile)
+	router.HandleFunc("/debug/pprof/symbol", pprof.Symbol)
+}
+
```
Comment (BerniVarga) [inline app.go]: The profiler part can be removed, it was added for test purposes only.

Comment (Nesze) [inline app.go]: I think it's fine to have it enabled by default. It's cheap and we can profile the apps instantly (probably should enable for the rest of the apps as well).

Comment (BerniVarga) [inline app.go]: Fine with me.

Comment (Zapadlo) [inline app.go]: +1 for having it enabled
